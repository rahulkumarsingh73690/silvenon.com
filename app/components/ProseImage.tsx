import CSS from 'csstype'
import cloudinary from '../utils/cloudinary'
import { screens } from '../consts'
import clsx from 'clsx'

// approximate px values of @tailwindcss/typography's 65ch on different viewports
const proseMaxWidth = {
  sm: '620px',
  lg: '700px',
  xl: '780px',
  '2xl': '920px',
}
const proseSizes: Array<keyof typeof proseMaxWidth> = ['sm', 'lg', 'xl', '2xl']
const imageSizes = [640, 768, 1024, 1536, 2048]
const largestImageSize = Math.max(...imageSizes)

export function getImageUrl(id: string, size: number) {
  return cloudinary(id, {
    crop: 'limit',
    width: size,
    format: 'auto',
    quality: 'auto',
  })
}

interface Props extends React.ImgHTMLAttributes<HTMLImageElement> {
  width: number
  height: number
  cloudinaryId?: string
  unsplash?: {
    name: string
    username: string
  }
}

export default function ProseImage({
  width,
  height,
  cloudinaryId,
  unsplash,
  ...props
}: Props) {
  const maxWidth = width < largestImageSize ? width / 2 : undefined
  const imgProps = { ...props, className: clsx(props.className, 'rounded-lg') }

  // eslint-disable-next-line jsx-a11y/alt-text
  let el = <img {...imgProps} />

  if (typeof cloudinaryId !== 'undefined') {
    el = (
      <div className="not-prose mx-auto" style={{ maxWidth }}>
        <div
          className="aspect-ratio"
          style={{
            '--width': width,
            '--height': height,
          }}
        >
          {/* eslint-disable-next-line jsx-a11y/alt-text */}
          <img
            {...imgProps}
            loading="lazy"
            src={getImageUrl(cloudinaryId, 1024)}
            srcSet={imageSizes
              .filter((size) => size < width)
              .concat(width)
              .map((size) => `${getImageUrl(cloudinaryId, size)} ${size}w`)
              .join(', ')}
            sizes={[
              ...proseSizes.map(
                (size) =>
                  `(min-width: ${screens[size]}) ${proseMaxWidth[size]}`,
              ),
              `calc(100vw - (var(--content-padding) * 2))`,
            ].join(', ')}
          />
        </div>
      </div>
    )
  }

  if (typeof unsplash !== 'undefined') {
    el = (
      <figure>
        {el}
        <figcaption>
          Photo by{' '}
          <a
            href={`https://unsplash.com/@${unsplash.username}?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText`}
          >
            {unsplash.username}
          </a>{' '}
          on{' '}
          <a href="https://unsplash.com/?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">
            Unsplash
          </a>
        </figcaption>
      </figure>
    )
  }

  return el
}

declare module 'react' {
  interface CSSProperties extends CSS.Properties<string | number> {
    '--width'?: number
    '--height'?: number
  }
}
