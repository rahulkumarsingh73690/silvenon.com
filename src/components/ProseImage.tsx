import React from 'react'
import CSS from 'csstype'
import cl from '../cloudinary'
import { screens } from '../consts'
import clsx from 'clsx'
import spacing from '../spacing.module.css'

// approximate px values of @tailwindcss/typography's 65ch on different viewports
const proseMaxWidth = {
  sm: '660px',
  lg: '740px',
  xl: '820px',
  '2xl': '1000px',
}
const proseSizes: Array<keyof typeof proseMaxWidth> = ['sm', 'lg', 'xl', '2xl']
const imageSizes = [640, 768, 1024, 1536, 2048]
const largestImageSize = Math.max(...imageSizes)

export function getImageUrl(id: string, size: number) {
  return cl(id, { crop: 'limit', width: size })
}

interface Props extends React.ImgHTMLAttributes<HTMLImageElement> {
  width: number
  height: number
  cloudinaryId?: string
  unsplash?: {
    name: string
    username: string
  }
}

export default function Image({
  width,
  height,
  cloudinaryId,
  unsplash,
  ...props
}: Props) {
  const maxWidth = width < largestImageSize ? width / 2 : undefined
  const imgProps = { ...props, className: clsx(props.className, 'rounded-lg') }

  // eslint-disable-next-line jsx-a11y/alt-text
  let el = <img {...imgProps} />

  if (typeof cloudinaryId !== 'undefined') {
    el = (
      <div
        className="ar mx-auto"
        style={{
          '--w': width,
          '--h': height,
          maxWidth,
        }}
      >
        <div className="ar-media">
          {/* eslint-disable-next-line jsx-a11y/alt-text */}
          <img
            {...imgProps}
            loading="lazy"
            src={getImageUrl(cloudinaryId, 1024)}
            srcSet={imageSizes
              .filter((size) => size < width)
              .concat(width)
              .map((size) => `${getImageUrl(cloudinaryId, size)} ${size}w`)
              .join(', ')}
            sizes={proseSizes
              .map(
                (size) =>
                  `(min-width: ${screens[size]}) ${proseMaxWidth[size]}`,
              )
              .concat(`calc(100vw - (${spacing.contentPadding} * 2))`)
              .join(', ')}
          />
        </div>
      </div>
    )
  }

  if (typeof unsplash !== 'undefined') {
    el = (
      <figure>
        {el}
        <figcaption>
          Photo by{' '}
          <a
            href={`https://unsplash.com/@${unsplash.username}?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText`}
          >
            {unsplash.username}
          </a>{' '}
          on{' '}
          <a href="https://unsplash.com/?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText">
            Unsplash
          </a>
        </figcaption>
      </figure>
    )
  }

  return el
}

declare module 'react' {
  interface CSSProperties extends CSS.Properties<string | number> {
    '--w'?: number
    '--h'?: number
  }
}
