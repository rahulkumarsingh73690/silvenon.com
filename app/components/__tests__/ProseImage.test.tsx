// @vitest-environment jsdom
import { render } from '@testing-library/react'
import ProseImage from '../ProseImage'

describe('ProseImage', () => {
  it('has correct output', () => {
    const { container } = render(
      <ProseImage
        alt="description"
        cloudinaryId="1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl"
        width={800}
        height={600}
      />,
    )
    expect(container.firstChild).toMatchSnapshot()
  })
})
