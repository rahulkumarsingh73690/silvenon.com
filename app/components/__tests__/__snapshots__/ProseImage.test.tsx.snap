// Vitest Snapshot v1

exports[`ProseImage > Cloudinary + Unsplash 1`] = `
<figure>
  <div
    class="not-prose mx-auto"
    style="max-width: 400px;"
  >
    <div
      class="aspect-ratio"
      style="--width: 800; --height: 600;"
    >
      <img
        alt="description"
        class="rounded-lg"
        loading="lazy"
        sizes="(min-width: 640px) 620px, (min-width: 1024px) 700px, (min-width: 1280px) 780px, (min-width: 1536px) 920px, calc(100vw - (var(--content-padding) * 2))"
        src="https://res.cloudinary.com/silvenon/image/upload/c_limit,w_1024,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl"
        srcset="https://res.cloudinary.com/silvenon/image/upload/c_limit,w_640,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl 640w, https://res.cloudinary.com/silvenon/image/upload/c_limit,w_768,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl 768w, https://res.cloudinary.com/silvenon/image/upload/c_limit,w_800,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl 800w"
      />
    </div>
  </div>
  <figcaption>
    Photo by
     
    <a
      href="https://unsplash.com/@silvenon?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText"
    >
      silvenon
    </a>
     
    on
     
    <a
      href="https://unsplash.com/?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText"
    >
      Unsplash
    </a>
  </figcaption>
</figure>
`;

exports[`ProseImage > Cloudinary 1`] = `
<div
  class="not-prose mx-auto"
  style="max-width: 400px;"
>
  <div
    class="aspect-ratio"
    style="--width: 800; --height: 600;"
  >
    <img
      alt="description"
      class="rounded-lg"
      loading="lazy"
      sizes="(min-width: 640px) 620px, (min-width: 1024px) 700px, (min-width: 1280px) 780px, (min-width: 1536px) 920px, calc(100vw - (var(--content-padding) * 2))"
      src="https://res.cloudinary.com/silvenon/image/upload/c_limit,w_1024,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl"
      srcset="https://res.cloudinary.com/silvenon/image/upload/c_limit,w_640,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl 640w, https://res.cloudinary.com/silvenon/image/upload/c_limit,w_768,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl 768w, https://res.cloudinary.com/silvenon/image/upload/c_limit,w_800,f_auto,q_auto/1_zs1MOwXPd-FcxoQ5RIMvyw_t3axyl 800w"
    />
  </div>
</div>
`;

exports[`ProseImage > Unsplash 1`] = `
<figure>
  <img
    alt="description"
    class="rounded-lg"
    src="/hello-world.jpg"
  />
  <figcaption>
    Photo by
     
    <a
      href="https://unsplash.com/@silvenon?utm_source=unsplash&amp;utm_medium=referral&amp;utm_content=creditCopyText"
    >
      silvenon
    </a>
     
    on
     
    <a
      href="https://unsplash.com/?utm_source=unsplash&utm_medium=referral&utm_content=creditCopyText"
    >
      Unsplash
    </a>
  </figcaption>
</figure>
`;

exports[`ProseImage > plain 1`] = `
<img
  alt="description"
  class="rounded-lg"
  src="/hello-world.jpg"
/>
`;
